### PR TITLE
COLLECT-472: Initial async task code to load instance of forms

### PR DESCRIPTION
**Pull request by nribeka**
This async task will read out the instances folder and do file path comparison to see which one already in the database and add those that are not in the database.

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/tasks/InstanceSyncTask.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (C) 2009 University of Washington
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/tasks/InstanceSyncTask.java]: You're the copyright holder and it's 2017!!

**Comment by lognaturel** [comment]
Thanks @nribeka! I still want to make sure that I understand what problem this solves. Is it mostly to make testing easier? @yanokwa you're the one who wanted this I think...

**Comment by yanokwa** [comment]
@lognaturel I've added more to issue #472!

File: collect_app/src/main/java/org/odk/collect/android/tasks/InstanceSyncTask.java (hunk 1 above)
Comment (nribeka) [reply collect_app/src/main/java/org/odk/collect/android/tasks/InstanceSyncTask.java]: So make it "2017 Nyoman Ribeka"? Yay ...

**Comment by nribeka** [comment]
@lognaturel & @yanokwa, this instance uploader ran pretty fast apparently. So I added the sync execution automatically on the onCreate of listing instances to be uploaded and listing of instances that can be edited.

**Comment by yanokwa** [comment]
@nribeka ✨Awesome✨ I haven't looked, but did you have some UI to tell the user that a scanning is happening? Also, I think we should also add this to Delete Saved Forms.

**Comment by nribeka** [comment]
I can add to the delete as well. Do just want a simple toast message when the scan start or you're thinking some more elaborate UI?

**Comment by yanokwa** [comment]
I was thinking the same UI as in the Fill Blank Form. Basically, show a message when you are scanning and update that message when the scanning stops.

device-2017-03-02-144414

File: collect_app/src/main/java/org/odk/collect/android/activities/FormChooserList.java [hunk 2]
```diff
@@ -151,7 +151,7 @@ protected void onResume() {
         super.onResume();
 
         if (mDiskSyncTask.getStatus() == AsyncTask.Status.FINISHED) {
-            SyncComplete(mDiskSyncTask.getStatusMessage());
+            syncComplete(mDiskSyncTask.getStatusMessage());
```
Comment (nribeka) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormChooserList.java:154]: This is part of the refactor.

**Comment by yanokwa** [comment]
When I try a submission of 1500 instances, I get the following...

```
03-06 10:54:32.196  3219  3219 I InstanceUploaderActiv: onCreate: creating
03-06 10:54:32.196  3219  3219 I InstanceUploaderActiv: onCreate: Beginning upload of 1500 instances!
03-06 10:54:32.206  3219 23199 V FA      : Using measurement service
03-06 10:54:32.206  3219 23199 V FA      : Connection attempt already in progress
03-06 10:54:32.212  3219  3219 I InstanceUploaderActiv: onResume: Resuming upload of 1500 instances!
03-06 10:54:32.213  3219 23199 V FA      : Using measurement service
03-06 10:54:32.213  3219 23199 V FA      : Connection attempt already in progress
03-06 10:54:32.213  3219 23199 V FA      : Activity resumed, time: 3242287
03-06 10:54:32.214  3219 23886 I PropertyManager: calling constructor
03-06 10:54:32.223  3219 23886 E SQLiteLog: (1) too many SQL variables
--------- beginning of crash
03-06 10:54:32.224  3219 23886 E AndroidRuntime: FATAL EXCEPTION: AsyncTask #31
03-06 10:54:32.224  3219 23886 E AndroidRuntime: Process: org.odk.collect.android, PID: 3219
03-06 10:54:32.224  3219 23886 E AndroidRuntime: java.lang.RuntimeException: An error occurred while executing doInBackground()
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.os.AsyncTask$3.done(AsyncTask.java:325)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:242)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:761)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: Caused by: android.database.sqlite.SQLiteException: too many SQL variables (code 1): , while compiling: SELECT canEditWhenComplete, deletedDate, status, jrVersion, jrFormId, instanceFilePath, date, submissionUri, displayName, _id, displaySubtext FROM instances WHERE (_id IN (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?))
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.database.sqlite.SQLiteConnection.nativePrepareStatement(Native Method)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.database.sqlite.SQLiteConnection.acquirePreparedStatement(SQLiteConnection.java:889)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.database.sqlite.SQLiteConnection.prepare(SQLiteConnection.java:500)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.database.sqlite.SQLiteSession.prepare(SQLiteSession.java:588)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.database.sqlite.SQLiteProgram.<init>(SQLiteProgram.java:58)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.database.sqlite.SQLiteQuery.<init>(SQLiteQuery.java:37)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.database.sqlite.SQLiteDirectCursorDriver.query(SQLiteDirectCursorDriver.java:44)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.database.sqlite.SQLiteDatabase.rawQueryWithFactory(SQLiteDatabase.java:1318)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.database.sqlite.SQLiteQueryBuilder.query(SQLiteQueryBuilder.java:399)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.database.sqlite.SQLiteQueryBuilder.query(SQLiteQueryBuilder.java:294)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at org.odk.collect.android.provider.InstanceProvider.query(InstanceProvider.java:165)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.content.ContentProvider.query(ContentProvider.java:1020)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.content.ContentProvider$Transport.query(ContentProvider.java:239)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.content.ContentResolver.query(ContentResolver.java:532)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.content.ContentResolver.query(ContentResolver.java:474)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at org.odk.collect.android.dao.InstancesDao.getInstancesCursor(InstancesDao.java:100)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at org.odk.collect.android.dao.InstancesDao.getInstancesCursor(InstancesDao.java:95)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at org.odk.collect.android.tasks.InstanceUploaderTask.doInBackground(InstanceUploaderTask.java:550)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at org.odk.collect.android.tasks.InstanceUploaderTask.doInBackground(InstanceUploaderTask.java:72)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at android.os.AsyncTask$2.call(AsyncTask.java:305)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
03-06 10:54:32.224  3219 23886 E AndroidRuntime: 	... 4 more
03-06 10:54:32.229  1297  1338 D gralloc_ranchu: gralloc_alloc: format 1 and usage 0x900 imply creation of host color buffer
03-06 10:54:32.231  1570  1819 W ActivityManager:   Force finishing activity org.odk.collect.android/.activities.InstanceUploaderActivity
03-06 10:54:32.234  1570  1819 W ActivityManager:   Force finishing activity org.odk.collect.android/.activities.InstanceUploaderList
03-06 10:54:32.235  1297  1338 D gralloc_ranchu: gralloc_alloc: format 1 and usage 0x900 imply creation of host color buffer
03-06 10:54:32.239  1297  1338 D gralloc_ranchu: gralloc_alloc: format 1 and usage 0x900 imply creation of host color buffer
03-06 10:54:32.242  3219  3249 E EGL_emulation: tid 3249: eglSurfaceAttrib(1174): error 0x3009 (EGL_BAD_MATCH)
03-06 10:54:32.242  3219  3249 W OpenGLRenderer: Failed to set EGL_SWAP_BEHAVIOR on surface 0x987e28e0, error=EGL_BAD_MATCH
03-06 10:54:32.259  1297  1338 D gralloc_ranchu: gralloc_alloc: format 1 and usage 0x900 imply creation of host color buffer
03-06 10:54:32.264  1297  1338 D gralloc_ranchu: gralloc_alloc: format 1 and usage 0x900 imply creation of host color buffer
03-06 10:54:32.267  1297  1966 E SurfaceFlinger: ro.sf.lcd_density must be defined as a build property
03-06 10:54:32.273  1297  1338 D gralloc_ranchu: gralloc_alloc: format 1 and usage 0x900 imply creation of host color buffer
03-06 10:54:32.285  3219  3249 E EGL_emulation: tid 3249: eglSurfaceAttrib(1174): error 0x3009 (EGL_BAD_MATCH)
03-06 10:54:32.285  3219  3249 W OpenGLRenderer: Failed to set EGL_SWAP_BEHAVIOR on surface 0x98709780, error=EGL_BAD_MATCH
03-06 10:54:32.287  3219  3219 I InstanceUploaderActiv: onPause: Pausing upload of 1500 instances!
03-06 10:54:32.287  3219 23199 D FA      : Connected to remote service
03-06 10:54:32.287  3219 23199 V FA      : Processing queued up service tasks: 3
03-06 10:54:32.287  3219 23199 V FA      : Screen exposed for less than 1000 ms. Event not sent. time: 75
03-06 10:54:32.288  3219 23199 V FA      : Activity paused, time: 3242362

**Review by yanokwa — CHANGES_REQUESTED** [review]
* No preferences for how it imports submissions. I thought we'd have saved vs finalized as an option.
* Scanning and scanning UI should be on `Edit Saved Form`,
* In Saved/Blank Forms, if I toggle the buttons I can see the UI differences. We should standardize on Blank Forms. 
* For consistency sake, make the scan language, `Finished scanning. %n forms added.`
* `Finished scanning the instances path` doesn't have the right coloring.
* If I don't have the blank form, but instead have lots of instances, it scans and says added, but it didn't actually add. 
* Is it possible to trigger a blank forms scan first, then do the instances scan.
* I would remove instance_scan_empty. If there is no instances folder, don't say anything.

**Comment by nribeka** [comment]
@yanokwa,

Updated except for the preference. You want the preference to be in the general settings, right?

**Comment by yanokwa** [comment]
@nribeka Good question. There is a refactor of the preferences due to be reviewed, so please file a separate issue to put an option for the scan import type as a preference.

File: collect_app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -480,4 +480,6 @@
 <string name="encrypted_form">Encrypted form</string>
 <string name="deleted_form">Deleted form</string>
 <string name="exit">Exit</string>
+<string name="instance_scan_completed">Finished scanning instances.</string>
```
Comment (yanokwa) [inline collect_app/src/main/res/values/strings.xml]: This should say `Finished scanning`. We have no mention of instances in the UI and I want to be consistent.

File: collect_app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -480,4 +480,6 @@
...
+<string name="instance_scan_count">%1$d instances added.</string>
```
Comment (yanokwa) [inline collect_app/src/main/res/values/strings.xml]: This should be `%1$d forms added`. Because consistency.

**Comment by yanokwa** [comment]
@nribeka Is it possible to trigger blank forms scan first, then do the instances scan? I think it makes for a better experience. That is, if you don't have the form, then you know you don't need to do the scan for that particular form.

**Comment by yanokwa** [comment]
@nribeka There is some jerkiness in the UI that I've filed at #530. Once we get this PR closed, it will be good for a quick win.

**Comment by nribeka** [comment]
@yanokwa we could start the form sync task before instance sync task. Do you want to do it here or as separate ticket? I'm worried we're trying to do too much in this PR 🤣

**Comment by yanokwa** [comment]
A separate PR is totally fine. File it please 🙏🏾

**Comment by nribeka** [comment]
@yanokwa we probably need to refactor that UI too to use slider / strip with fragments instead of TabHost. The fragments then can share the same layout.

Currently they are on the same tab, but using different layout. One is form_manage_list.xml, the other is data_manage_list.xml. Both layout file have the exact same elements.

**Comment by yanokwa** [comment]
@nribeka This is looking good to me. @lognaturel I think this is ready for your review.

**Comment by lognaturel** [comment]
@yanokwa You need to have faith in your review! ;)

**Pull request merged**